### ISSUE TITLE
[ASTextNode] Additional attributes for links, such as a custom color, aren't applied.

### ISSUE THREAD

**Issue by jarodl** [facebook/AsyncDisplayKit#571]
When adding links to an attributed string, the value of `NSForegroundColorAttributeName` is not respected. The default blue color is used. Is there anyway to get around this? Setting the underline and font keys work fine though. Any ideas?

**Comment by mmiedlarz** [comment]
Got same problem. textNode.tintColor does not work at all too.

**Comment by appleguy** [comment]
@jarodl @mmiedlarz there is a working implementation in examples/Kittens, in the BlurbNode.m file:

"  [string addAttributes:@{
                          kLinkAttributeName: [NSURL URLWithString:@"http://placekitten.com/"],
                          NSForegroundColorAttributeName: [UIColor grayColor],
                          NSUnderlineStyleAttributeName: @(NSUnderlineStyleSingle | NSUnderlinePatternDot),
                          }
"

@mmiedlarz, I'd like to fix the tintColor issue.  What is the behavior you expect from tintColor?  I haven't used it extensively with UILabel, for example.  How should it affect the text color, link color, etc; I assume it should not override any explicitly-set colors, but serve as the foreground text color when nothing has been set?

**Comment by mmiedlarz** [comment]
@appleguy Sorry for late. I'm in the middle of vacation now :)
So it looks like this feature is broken on 1.2.3 version. I've prepared demo here:
https://github.com/mmiedlarz/ADKBrokenURLColorDemo

And about tintColor  - if you set it on textView, it will override all colors which you set on attributed string (see demo).

Cheers.

**Comment by mmiedlarz** [comment]
Hey,

Any progress for this issue? Maybe in 2.0?

**Comment by appleguy** [comment]
@mmiedlarz thanks for pinging.  Is the only issue you're seeing related to tintColor?  I mentioned the apparently-working Kittens example, so I'm not too sure if there's something besides tintColor that you still see as broken in master.

**Comment by mmiedlarz** [comment]
@appleguy I can't build Kittens example (I think that its not compatible with 1.2.3 version), 
but its reproduced in my demo ( https://github.com/mmiedlarz/ADKBrokenURLColorDemo ). 
TextNode just ignore color set by:
```
NSDataDetector *detect = [[NSDataDetector alloc] initWithTypes:NSTextCheckingTypeLink error:nil];
NSArray *matches = [detect matchesInString:attrString.string options:0 range:NSMakeRange(0, [attrString length])];
NSMutableAttributedString *str = [[NSMutableAttributedString alloc] initWithAttributedString:attrString];
for(NSTextCheckingResult *result in matches)
{
    [str addAttribute:NSLinkAttributeName value:result.URL range:result.range];
    [str addAttribute:NSForegroundColorAttributeName value:[UIColor redColor] range:result.range];
    [str addAttribute:NSUnderlineStyleAttributeName value:@(NSUnderlineStyleSingle | NSUnderlinePatternDot) range:result.range];
}
```
If I remove NSLinkAttributeName attriubte, ForegroundColor works fine.

**Comment by RuiAAPeres** [comment]
The tint color for me doesn't work either, I would expect the link's color to change as well.

**Comment by appleguy** [comment]
Hmm, this is a good bug.  We just fixed some attribute handling in ASTextNode, but if I'm understanding the test project correctly, this one is still not fixed.  Sorry about that guys — I'm not sure what's going on yet, especially since Paper and Instant Articles both clearly use non-default link colors and most certainly uses ASTextNode.

@mmiedlarz just want to say again how much I appreciate you uploading the test project.  It is very helpful to getting something like this fixed.  I'm sorry there's been a long delay on this, but as you can see there has been a lot of activity on the project with really big efforts, like adopting a big update to the text rendering engine and important internal capabilities like .interfaceState and .hierarchyState.  We are now starting to get to issues like this.

@samhsiung - I don't expect you to spend a lot of time looking at this, but since you recently figured out some really quite tricky issues with ASTextNode and attribute handling, do you think this bug might stand out to you?  There is definitely special handling of NSLinkAttributeName in the code, and that's probably to blame, considering that removing that attribute from the sample code provided causes the color to come through correctly.

**Comment by appleguy** [comment]
I am hopeful we can get this resolved much sooner than 2.1, but the 2.0 milestone is already very large and I am not confident it will be fixed in time.  Putting it in 2.1 for now, but we will pull it forward as soon as it is fixed (even if that's a 1.9.5 release).

**Comment by mmiedlarz** [comment]
@appleguy I updated TestProject to 1.9.3, but issue still existed. But I found, that if you use custom link attributes in NSAttributedString and ASTextNode, everything works fine.

**Comment by samhsiung** [comment]
In the project example, it looks like `UITextView` has the same issues. `NSForegroundColorAttributeName` isn't respected if `NSLinkAttributeName` is set. However `UITextView` has a `linkTextAttributes` property which will allow you to customize foreground color of links like:

`textView.linkTextAttributes = @{NSForegroundColorAttributeName:[UIColor redColor]};`

http://stackoverflow.com/questions/25457131/setting-nslinkattributename-font-color

It's probably possible to support such a feature like this since `TTTAttributedLabel` seems to as well, but if this is correct, then this should be a feature we would have to add and not a bug.

**Comment by samhsiung** [comment]
Logged this feature request for adding `linkTextAttributes` to `ASTextNode`: https://github.com/facebook/AsyncDisplayKit/issues/967

**Comment by nguyenhuy** [comment]
Looks like this is fixed by #1127. Closing...
